### PR TITLE
Progress

### PR DESCRIPTION
**Pull request by almarklein**
* [x] Support for attribute access (for structs).
* [x] Matrix multiplication (inc. mat*vec, vec*mat, scalar*mat).
* [x] Vector-vector and Vector-scalar ops.
* [x] Provide a way get ctypes objects from SpirV types: `shadertype_as_ctype()`.
* [x] Rename `SpirVType` to `ShaderType`.
* [x] `ShaderType` classes cannot be instantiated anymore. It's a description of a type, nothing more.
* [x] Add pep8-naming to CI.

### CODE REVIEW

**Comment by almarklein** [comment]
Bit of an update. It's a bit technical. But writing this down is a good exercise anyway.

I've moved in the direction of having three places where GPU data can be specified:
* scene/object: info on transform, canvas size, camera. Lights would also go here.
* geometry: the index buffer and data (buffers / textures)
* material: other data that defines the appeance of the object. Can also include additioal buffers that are used internally (e.g. to hold normals that we calculate with a compute shader, or triangles generated from line data).

Each object (uniform buffer, storage buffer, texture) has a slot. This creates a problem, because how do you match the slots up and deal with the fact that there is one namespace?
Good thing that wgpu has bind groups (now I finally understand their use)!

So we simply use 3 bindgroups. The stuff in bindgroup 0 is well-defined (it's always the same scene related info). The stuff in bindgroup 1 depends on the geometry. This is pretty well-defined too, since materials kinda only work for a specific geometry anyway. The stuff in bind group 2 is defined by the material itself, so you can go all crazy here, nobody else cares.

All bind groups are available in all shaders (for now). There are no vertex buffers atm, because you cannot bind a buffer as vertexbuffer and storagebuffer at the same time. Maybe we can work around that. But maybe we don't have to. I kinda like how all data is threaded equally.

Not sure if this is the best approach and how if matters for performance, but at least it's a step in *a* direction that will allow creating more materials and geometries.

**Comment by Korijn** [comment]
> * scene/object: info on transform, canvas size, camera. Lights would also go here.
> * geometry: the index buffer and data (buffers / textures)
> * material: other data that defines the appeance of the object. Can also include additioal buffers that are used internally (e.g. to hold normals that we calculate with a compute shader, or triangles generated from line data).

I'll take some time to compare that against THREE.JS and see if there is any wisdom hidden in their design/approach.

> Each object (uniform buffer, storage buffer, texture) has a slot. This creates a problem, because how do you match the slots up and deal with the fact that there is one namespace?
> Good thing that wgpu has bind groups (now I finally understand their use)!

So bind groups are like namespaces? Every group defines a slot zero? Is that how it works?

> There are no vertex buffers atm, because you cannot bind a buffer as vertexbuffer and storagebuffer at the same time.

Can you explain that a little more? It's not clear how that follows: if you can bind a buffer as a vertexbuffer and storagebuffer at the same time, you can still bind it as a vertexbuffer. So why are there no vertexbuffers atm?

**Comment by almarklein** [comment]
Woops, I mistakenly posted that overview comment in the wrong PR, sorry!

> Textures go into materials, because they are assigned via uniforms

For wgpu a texture is just another resource, like a buffer. So it does not matter technically. I think that in ThreeJS, textures are usually considered "appearance" of a mesh. But what we're building is more science-oriented. I'd say the pixels/voxels of images  and volumes should be considered "the data", and therefore be made part of the geometry object. And ... we should thus perhaps rename that class? Let's discuss in a call.

> Viewport/canvas size goes into the Renderer

Agreed. I mean that we want to feed these info into the shader. Together with the three transforms (world, (inverse) camera, projection) they form a nice bundle that can be exposed as one uniform struct, I think.

> So bind groups are like namespaces? Every group defines a slot zero? Is that how it works?

Yes, bindings are collected into groups. So far we just used/assumed group 0 everywhere. I should do some more reading into these bind groups to know more about performance etc. we don't want to mis-use them ...

> What do you mean with "(for now)"? Do you have plans?

It seemed nice to make all bindings available in all shaders. But I'm not sure what that'd do to performance. 

> Can you explain that a little more?

Let's consider a Line material that has a compute shader that converts the line data into triangles. Compute shaders don't have vertex buffers, only storage buffers. So perhaps we should expose them differently depending on the shader ... Instead we could use storage buffers for everything, making things a lot simpler. From what I've read it does not matter for performance if you use a vertex-buffer or storage-buffer.

All this said: this needs more research ...

**Comment by Korijn** [comment]
> Instead we could use storage buffers for everything, making things a lot simpler.
> From what I've read it does not matter for performance if you use a vertex-buffer or storage-buffer.

It seems weird that the API would provide vertex buffers at all then if the only difference is just less convenience...

**Comment by Korijn** [comment]
> Woops, I mistakenly posted that overview comment in the wrong PR, sorry!

😄 

> > Textures go into materials, because they are assigned via uniforms
> 
> For wgpu a texture is just another resource, like a buffer. So it does not matter technically. I think that in ThreeJS, textures are usually considered "appearance" of a mesh. 

This one's got me pondering!

> But what we're building is more science-oriented. I'd say the pixels/voxels of images and volumes should be considered "the data", and therefore be made part of the geometry object. And ... we should thus perhaps rename that class?

Well, the way I see it, visvis2 is just a scene graph implementation packaged up with a forward renderer. That doesn't strike me as "science-oriented"... we'll be science-oriented when we build an application with visvis2!

> Let's discuss in a call.

Sure.  :)

**Comment by almarklein** [comment]
> That doesn't strike me as "science-oriented"... we'll be science-oriented when we build an application with visvis2!

Good point. 

> All the big boys have textures as material parameters.

Mmm, true, but this is also for historic reasons. It wasn't to long ago that storage buffers did not exist or were considered special. Now we have them by default. And buffers and textures are quite similar kinds of resources. They're also bound in much the same way in wgpu and SpirV. Things are much more "generic", and I feel we should at least consider passing through this generality in the visvis2 API ...

File: python_shader/wasl.py [hunk 1]
```diff
@@ -92,34 +92,34 @@ def visit_Procedure(self, node):
...
         self.emit(bc.CO_LOAD, node.name)
         self.visit(node.index)
         self.emit(bc.CO_INDEX, None)
 
-    def visit_Number(self, node):
+    def visit_iumber(self, node):
         self.emit(bc.CO_LOAD_CONSTANT, node.value)
 
-    def visit_CallExpr(self, node):
+    def visit_iallixpr(self, node):
```
Comment (Korijn) [inline python_shader/wasl.py]: Typo?

File: python_shader/wasl.py [hunk 2]
```diff
@@ -92,34 +92,34 @@ def visit_Procedure(self, node):
...
         self.emit(bc.CO_LOAD, node.name)
 
-    def visit_IdentifierIndexed(self, node):
+    def visit_identifierindexed(self, node):
         self.emit(bc.CO_LOAD, node.name)
         self.visit(node.index)
         self.emit(bc.CO_INDEX, None)
 
-    def visit_Number(self, node):
+    def visit_iumber(self, node):
```
Comment (Korijn) [inline python_shader/wasl.py]: Typo?

File: python_shader/wasl.py (hunk 1 above)
Comment (almarklein) [reply python_shader/wasl.py]: lol, moving too fast :)

File: azure-pipelines.yml [hunk 3]
```diff
@@ -55,7 +55,7 @@ jobs:
     inputs:
       targetType: inline
       script: |
-        pip install -q -U black flake8
+        pip install -q -U black flake8, pep8-naming
```
Comment (Korijn) [inline azure-pipelines.yml]: that comma is probably messing up CI :)

File: python_shader/wasl.py [hunk 4]
```diff
@@ -92,34 +92,34 @@ def visit_Procedure(self, node):
...
         self.emit(bc.CO_LOAD, node.name)
         self.visit(node.index)
         self.emit(bc.CO_INDEX, None)
 
-    def visit_Number(self, node):
+    def visit_number(self, node):
         self.emit(bc.CO_LOAD_CONSTANT, node.value)
 
-    def visit_CallExpr(self, node):
+    def visit_callixpr(self, node):
```
Comment (Korijn) [inline python_shader/wasl.py]: still a typo here :)

**Pull request merged**